### PR TITLE
[Feature] Enable Ruby Proc for `require_confirm_worker_params`

### PR DESCRIPTION
**Pull request by khoaanguyenn**
# Background

- We need a feature to enable `require_confirm_worker_params` to also accept[ Ruby callable object](https://livebook.manning.com/book/the-well-grounded-rubyist-second-edition/chapter-14/36) in the configuration as follows

```ruby
SidekiqAdhocJob.configure do |config|
  config.module_names = [:Module]

  config.require_confirm_worker_names = %w[TestWorker TestJob] # current implementation
  config.require_confirm_worker_names = ->(_worker_name) { true } # apply to all workers
  config.require_confirm_worker_names = RubyCallable.new # Ruby callable object 
  ...
end
SidekiqAdhocJob.init
```

- From the PR #30 , confirmation feature was added. However, when input incorrect challenge keyword, nothing happens but closing the prompt. This is not a good user experience, because they might not know that the inputed keyword is incorrect.

- The [Sidekiq upgrade 6.1.3 -> 6.2.0](https://github.com/mperham/sidekiq/issues/4850), causes `Sidekiq::Web needs a valid Rack session for CSRF protection.` to every requests to server, thus [requests test suite](https://github.com/gohkhoonhiang/sidekiq_adhoc_job/tree/master/spec/sidekiq_adhoc_job/requests/jobs) eventually fail and print out the following error message
```
6) GET /adhoc_jobs/:name job not found redirects to index page
     Failure/Error: get '/adhoc-jobs/invalid_worker'

     RuntimeError:
       Sidekiq::Web needs a valid Rack session for CSRF protection. If this is a Rails app,
       make sure you mount Sidekiq::Web *inside* your application routes:


       Rails.application.routes.draw do
         mount Sidekiq::Web => "/sidekiq"
         ....
       end


       If this is a Rails app in API mode, you need to enable sessions.

         https://guides.rubyonrails.org/api_app.html#using-session-middlewares

       If this is a bare Rack app, use a session middleware before Sidekiq::Web:

         # first, use IRB to create a shared secret key for sessions and commit it
         require 'securerandom'; File.open(".session.key", "w") {|f| f.write(SecureRandom.hex(32)) }

         # now use the secret with a session cookie middleware
         use Rack::Session::Cookie, secret: File.read(".session.key"), same_site: true, max_age: 86400
         run Sidekiq::Web
     # ./spec/sidekiq_adhoc_job/requests/jobs/show_spec.rb:130:in `block (3 levels) in <top (required)>'
```
- The changes in UI from PR #29 also breaks the [`show_spec`](https://github.com/gohkhoonhiang/sidekiq_adhoc_job/blob/master/spec/sidekiq_adhoc_job/requests/jobs/show_spec.rb#L19), since it added `id` attribute for the corresponding `<form>` to attach `submit` event on top of it.
# Design

- Adjust javascript confirmation.
- Allows `#require_confirm_worker_params` to accept Ruby callable object.
- Add `Gemfile.lock` to this repository.
- Add more helpers to shared contexts.
- Disable Sidekiq CSRF in the test suite. [Reference](https://github.com/mperham/sidekiq/blob/09f2a4ca3bbb8439b745d4f81a241815d28ffe66/lib/sidekiq/web/csrf_protection.rb#L41)
    ```ruby
    def disable_csrf
      allow_any_instance_of(Sidekiq::Web::CsrfProtection).to receive(:accept?).and_return(true)
    end
    ```
- Fix all failed specs.
- Refactor specs
# Impact

- Show alert when the inputed challenge keyword is incorrect.
- Enable `#require_confirm_worker_params` to accept Ruby callable object such as Proc, lambda or any Ruby class that implements `#call` method.
- All tests are passed ✅ 
# Caveats

None.
# Testing

Install this branch to test the new behaviors.
```bash
  gem 'sidekiq_adhoc_job', git: 'https://github.com/gohkhoonhiang/sidekiq_adhoc_job.git', branch: 'feature/require_confirm_proc'
```

1. Ensure the test suite all passed
    ```bash
      bundle exec rspec
    ```
2. When inputting incorrect challenge keyword the alert `Your confirmation input is incorrect, try again.` should be displayed.
3. Can config Ruby callable object to `#require_confirm_worker_params`.
# Docs

- Ruby callable object. [The Well-Grounded Rubyist, chapter 14](https://livebook.manning.com/book/the-well-grounded-rubyist-second-edition/chapter-14/36)
- Sidekiq CSRF protection middleware. [Reference](https://github.com/mperham/sidekiq/blob/09f2a4ca3bbb8439b745d4f81a241815d28ffe66/lib/sidekiq/web/csrf_protection.rb#L41)
- Sidekiq CSRF protection test case. [Reference](https://github.com/mperham/sidekiq/blob/09f2a4ca3bbb8439b745d4f81a241815d28ffe66/test/csrf_test.rb#L63)

### CODE REVIEW

**Comment by teohm** [comment]
The `require_confirm_worker_params` changes look good to me.

However, it looks like the CI build failed?

**Comment by khoaanguyenn** [comment]
Yes, the `bundler` gem now only supports `x86_64-darwin-21` platform, but the current platform is `x86_64-linux` one. [Travis CI build](https://app.travis-ci.com/github/gohkhoonhiang/sidekiq_adhoc_job/jobs/593450124). Thus, we need to change the OS build config

**Comment by gohkhoonhiang** [comment]
@khoaanguyenn Can we not commit the lock file and try, because you specified `x86_64-darwin-21` in the lock file? Also, what was the reason to commit the lock file?

File: spec/sidekiq_adhoc_job_spec.rb [hunk 1]
```diff
@@ -19,29 +21,66 @@
...
-  end
 
-  describe '.config' do
-    it 'returns configuration object' do
-      subject.configure { |config| config.module_names = [:'SidekiqAdhocJob::Test'] }
-      expect(subject.config).to be_an_instance_of(SidekiqAdhocJob::Configuration)
+    context '#require_confirm_worker_names' do
+      let(:config) { subject.config }
+
+      context 'When list of String' do
```
Comment (gohkhoonhiang) [inline spec/sidekiq_adhoc_job_spec.rb:35]:
```suggestion
      context 'when given list of Strings' do
```

File: spec/sidekiq_adhoc_job_spec.rb [hunk 2]
```diff
@@ -19,29 +21,66 @@
...
+        let(:require_confirm_worker_names) { %w[Test::IncludedWorker] }
+
+        it 'returns all matching worker names' do
+          expect(config.require_confirmation?('Test::IncludedWorker')).to eq true
+          expect(config.require_confirmation?('Test::NotIncludedWorker')).to eq false
+        end
+      end
+
+      context 'When list of Symbols' do
```
Comment (gohkhoonhiang) [inline spec/sidekiq_adhoc_job_spec.rb:44]:
```suggestion
      context 'when given list of Symbols' do
```

File: spec/sidekiq_adhoc_job_spec.rb [hunk 3]
```diff
@@ -19,29 +21,66 @@
...
+        let(:require_confirm_worker_names) { %i[Test::IncludedWorker] }
+
+        it 'returns all matching worker names' do
+          expect(config.require_confirmation?('Test::IncludedWorker')).to eq true
+          expect(config.require_confirmation?('Test::NotIncludedWorker')).to eq false
+        end
+      end
+
+      context 'When a Proc' do
```
Comment (gohkhoonhiang) [inline spec/sidekiq_adhoc_job_spec.rb:53]:
```suggestion
      context 'when given a Proc' do
```

File: spec/sidekiq_adhoc_job_spec.rb [hunk 4]
```diff
@@ -19,29 +21,66 @@
...
+        let(:require_confirm_worker_names) { ->(worker_name) { worker_name.start_with?('Test') } }
+
+        it 'returns all matching worker names' do
+          expect(config.require_confirmation?('Test::DummyWorker')).to eq(true)
+          expect(config.require_confirmation?('NoTest::DummyWorker')).to eq(false)
+        end
+      end
+
+      context 'when a Class implements #call' do
```
Comment (gohkhoonhiang) [inline spec/sidekiq_adhoc_job_spec.rb:62]:
```suggestion
      context 'when given a Class that implements #call' do
```

**Review by gohkhoonhiang — COMMENTED** [review]
Thanks for the contribution, and fixing some styling issues! There're just some texts to fix in the specs. Could you also help to update the README https://github.com/gohkhoonhiang/sidekiq_adhoc_job/blob/master/README.md#usage section to include the various ways to configure the `require_confirm_worker_params`?

File: README.md [hunk 5]
```diff
@@ -23,6 +23,11 @@ SidekiqAdhocJob.configure do |config|
   config.module_names = ['YourProject::Worker']
   config.strategy_name = :active_job
   config.load_paths = ['app/jobs/**/*.rb']
+
+  config.require_confirm_worker_names = ['YourProject::Worker::A', 'YourProject::Worker::B']
+  # List of Symbols                   = %i[YourProject::Worker::A YourProject::Worker::B]
+  # Apply to worker name end with "A" = ->(worker_class) { worker_class.end_with?('A') }
+  # Ruby class that implements #call  = CallableObject.new
```
Comment (gohkhoonhiang) [inline README.md:30]:
```suggestion
  config.require_confirm_worker_names     = %w[YourProject::Worker::A YourProject::Worker::B]
  # List of Symbols                       = %i[YourProject::Worker::A YourProject::Worker::B]
  # Apply to workers name ending with "A" = ->(worker_class) { worker_class.end_with?('A') }
  # Ruby class that implements #call      = CallableObject.new
```

File: README.md [hunk 6]
```diff
@@ -35,7 +40,7 @@ Options:
   - `active_job`: check for all classes that extend `ActiveJob::Base`
   - `rails_application_job`: check for all classes that extend `ApplicationJob`
 - `load_paths` (optional - default `[]`): takes in a list of file paths that the gem should load when initializing, in order to include the necessary classes in the app `ObjectSpace`
-- `require_confirm_worker_names` (optional - default `[]`): takes in a list of fully namespaced worker class names that the web UI will request for confirmation before running the job
+- `require_confirm_worker_names` (optional - default `[]`): takes in a Ruby callable object as predicate or fully namespaced worker class names that require confirmation before running the job through web UI.
```
Comment (gohkhoonhiang) [inline README.md:43]:
```suggestion
- `require_confirm_worker_names` (optional - default `[]`): takes in a Ruby callable object, or a Proc, or fully namespaced worker class names that require confirmation before running the job through web UI.
```

File: spec/sidekiq_adhoc_job_spec.rb [hunk 7]
```diff
@@ -19,29 +21,66 @@
...
-  end
 
-  describe '.config' do
-    it 'returns configuration object' do
-      subject.configure { |config| config.module_names = [:'SidekiqAdhocJob::Test'] }
-      expect(subject.config).to be_an_instance_of(SidekiqAdhocJob::Configuration)
+    context '#require_confirm_worker_names' do
+      let(:config) { subject.config }
+
+      context 'When given list of Strings' do
```
Comment (gohkhoonhiang) [inline spec/sidekiq_adhoc_job_spec.rb:35]:
```suggestion
      context 'when given list of Strings' do
```

File: spec/sidekiq_adhoc_job_spec.rb [hunk 8]
```diff
@@ -19,29 +21,66 @@
...
+        let(:require_confirm_worker_names) { %w[Test::IncludedWorker] }
+
+        it 'returns all matching worker names' do
+          expect(config.require_confirmation?('Test::IncludedWorker')).to eq true
+          expect(config.require_confirmation?('Test::NotIncludedWorker')).to eq false
+        end
+      end
+
+      context 'When given list of Symbols' do
```
Comment (gohkhoonhiang) [inline spec/sidekiq_adhoc_job_spec.rb:44]:
```suggestion
      context 'when given list of Symbols' do
```

File: spec/sidekiq_adhoc_job_spec.rb [hunk 9]
```diff
@@ -19,29 +21,66 @@
...
+        let(:require_confirm_worker_names) { %i[Test::IncludedWorker] }
+
+        it 'returns all matching worker names' do
+          expect(config.require_confirmation?('Test::IncludedWorker')).to eq true
+          expect(config.require_confirmation?('Test::NotIncludedWorker')).to eq false
+        end
+      end
+
+      context 'When given a Proc' do
```
Comment (gohkhoonhiang) [inline spec/sidekiq_adhoc_job_spec.rb:53]:
```suggestion
      context 'when given a Proc' do
```

File: spec/sidekiq_adhoc_job_spec.rb [hunk 10]
```diff
@@ -19,29 +21,66 @@
...
+        let(:require_confirm_worker_names) { ->(worker_name) { worker_name.start_with?('Test') } }
+
+        it 'returns all matching worker names' do
+          expect(config.require_confirmation?('Test::DummyWorker')).to eq(true)
+          expect(config.require_confirmation?('NoTest::DummyWorker')).to eq(false)
+        end
+      end
+
+      context 'When given a Class that implements #call' do
```
Comment (gohkhoonhiang) [inline spec/sidekiq_adhoc_job_spec.rb:62]:
```suggestion
      context 'when given a Class that implements #call' do
```

**Review by gohkhoonhiang — APPROVED** [review]

**Pull request merged**